src/System.Text.Utf8String/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,782 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Text.Utf16;
+
+namespace System.Text.Utf8
+{
+    [DebuggerDisplay("{ToString()}u8")]
+    public partial ref struct Utf8Span
+    {
+        private readonly ReadOnlySpan<byte> _buffer;
+
+        private const int StringNotFound = -1;
+
+        static Utf8Span s_empty => default;
+
+        // TODO: Validate constructors, When should we copy? When should we just use the underlying array?
+        // TODO: Should we be immutable/readonly?
+        public Utf8Span(ReadOnlySpan<byte> buffer)
+        {
+            _buffer = buffer;
+        }
+
+        public Utf8Span(byte[] utf8bytes)
+        {
+            _buffer = new ReadOnlySpan<byte>(utf8bytes);
+        }
+
+        public Utf8Span(byte[] utf8bytes, int index, int length)
+        {
+            _buffer = new ReadOnlySpan<byte>(utf8bytes, index, length);
+        }
+
+        public Utf8Span(string s)
+        {
+            if (s == null)
+            {
+                throw new ArgumentNullException("s", "String cannot be null");
+            }
+
+            if (s == string.Empty)
+            {
+                _buffer = ReadOnlySpan<byte>.Empty;
+            }
+            else
+            {
+                _buffer = new ReadOnlySpan<byte>(GetUtf8BytesFromString(s));
+            }
+        }
+
+        /// <summary>
+        /// This constructor is for use by the compiler.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Utf8Span(RuntimeFieldHandle utf8Data, int length) : this(CreateArrayFromFieldHandle(utf8Data, length))
```

Comment:
Two questions:
1. Do we need this ctor on Utf8Span, or we just need it on Utf8String?
2. Is the ctor even useful for the compiler? i.e. will the compiler benefit from it or it will just create byte[] and pass into the byte[] ctor?

@VSadov, @jaredpar

Reply:
Let's leave it for now, since it exists in Utf8String.